Src/Particle/AMReX_WriteBinaryParticleData.H

```diff
@@ -0,0 +1,688 @@
+            if ( ! LevelDir.empty() && LevelDir[LevelDir.size()-1] != '/') LevelDir += '/';
+
+            LevelDir = amrex::Concatenate(LevelDir + "Level_", lev, 1);
+
+            if (MyProc == IOProcNumber)
+            {
+                if ( ! amrex::UtilCreateDirectory(LevelDir, 0755))
+                {
+                    amrex::CreateDirectoryFailed(LevelDir);
+                }
+            }
+            ParallelDescriptor::Barrier();
+        }
+
+        // Write out the header for each particle
+        if (gotsome and (MyProc == IOProcNumber))
+        {
+            std::string HeaderFileName = LevelDir;
+            HeaderFileName += "/Particle_H";
+            std::ofstream ParticleHeader(HeaderFileName);
+
+            pc.ParticleBoxArray(lev).writeOn(ParticleHeader);
+            ParticleHeader << '\n';
+
+            ParticleHeader.flush();
+            ParticleHeader.close();
+        }
+    }
+
+    int maxnextid = PC::ParticleType::NextID();
+    ParallelDescriptor::ReduceIntMax(maxnextid, IOProcNumber);
+
+    Vector<int64_t> np_on_rank(NProcs,-1L);
+    for (int lev = 0; lev <= pc.finestLevel(); lev++)
+    {
+        for (MFIter mfi = pc.MakeMFIter(lev, false); mfi.isValid(); ++mfi)
+        {
+            int rank = pc.ParticleDistributionMap(lev)[mfi.index()];
+            np_on_rank[rank] += np_per_grid[lev][mfi.index()];
+        }
+    }
+
+    using ParticleReal = typename PC::RealType;
+    std::size_t psize = PSizeInFile<ParticleReal>(write_real_comp, write_int_comp);
+    Vector<int64_t> rank_start_offset(NProcs);
+    for (int ip = 0; ip < NProcs; ++ip)
+    {
+        auto info = AsyncOut::GetWriteInfo(ip);
+        rank_start_offset[ip] = (info.ispot == 0) ? 0 : rank_start_offset[ip-1] + np_on_rank[ip-1]*psize;
+    }
+
+    // make tmp particle tiles in pinned memory to write
+    using PinnedPTile = ParticleTile<NStructReal, NStructInt, NArrayReal, NArrayInt,
+                                     PinnedArenaAllocator>;
+    auto myptiles = std::make_shared<Vector<std::map<std::pair<int, int>,PinnedPTile> > >();
+    myptiles->resize(pc.finestLevel()+1);
+    for (int lev = 0; lev <= pc.finestLevel(); lev++)
+    {
+        for (MFIter mfi = pc.MakeMFIter(lev); mfi.isValid(); ++mfi)
+        {
+            auto& new_ptile = (*myptiles)[lev][std::make_pair(mfi.index(),
+                                                              mfi.LocalTileIndex())];
+
+            if (np_per_grid[lev][mfi.index()] > 0)
+            {
+                const auto& ptile = pc.ParticlesAt(lev, mfi);
+                new_ptile.resize(ptile.numParticles());
+                amrex::copyParticles(new_ptile, ptile);
+            }
+        }
+    }
+
+    auto wrc = std::make_shared<Vector<int> >(write_real_comp);
+    auto wic = std::make_shared<Vector<int> >(write_int_comp);
+    auto rcn = std::make_shared<Vector<std::string> >(real_comp_names);
+    auto icn = std::make_shared<Vector<std::string> >(int_comp_names);
```

Comment:
Are they needed?  Can we just use `write_real_comp` etc. directly in the lambda below?

Reply:
Indeed they are not. Fixed.